src/components/ColorPicker.tsx

```diff
@@ -5,46 +5,58 @@ import Saturation from "./Saturation";
 
 import styles from "../styles.css";
 import { ColorModel, HSV, ColorPickerBaseProps, AnyColor } from "../types";
-import { equalColorObjects } from "../utils/compare";
 import formatClassName from "../utils/formatClassName";
 
 interface Props<T extends AnyColor> extends Partial<ColorPickerBaseProps<T>> {
   colorModel: ColorModel<T>;
 }
 
+function useEventCallback<T>(event?: (arg: T) => void) {
+  const eventRef = useRef(event);
+
+  useEffect(() => {
+    eventRef.current = event;
+  });
+
+  return useCallback((arg: T) => eventRef.current && eventRef.current(arg), []);
+}
+
 const ColorPicker = <T extends AnyColor>({
   className = "",
   colorModel,
   color = colorModel.defaultColor,
   onChange,
 }: Props<T>) => {
+  // Save onChange callback in the ref for avoiding "useCallback hell"
+  const onChangeRef = useEventCallback<T>(onChange);
+
   // No matter which color model is used (HEX, RGB or HSL),
   // all internal calculations are based on HSV model
   const [hsv, updateHsv] = useState<HSV>(() => colorModel.toHsv(color));
 
   // By using this ref we're able to prevent extra updates
   // and the effects recursion during the color conversion
-  const cache = useRef({ color, hsv });
+  const cache = useRef(color);
 
   // Update local HSV if `color` property value is changed,
   // but only if that's not the same color that we just sent to the parent
   useEffect(() => {
-    if (!colorModel.equal(color, cache.current.color)) {
+    if (!colorModel.equal(color, cache.current)) {
       const newHsv = colorModel.toHsv(color);
-      cache.current = { hsv: newHsv, color };
+      cache.current = color;
       updateHsv(newHsv);
     }
   }, [color, colorModel]);
 
-  // If HSV is changed, convert it to the output format, send to the parent component
+  // Сonvert HSV to the output format, if it is changed send to the parent component
   // and save the new color to the ref to prevent unnecessary updates
   useEffect(() => {
-    if (!equalColorObjects(hsv, cache.current.hsv)) {
-      const newColor = colorModel.fromHsv(hsv);
-      cache.current = { hsv, color: newColor };
-      if (onChange) onChange(newColor);
+    const newColor = colorModel.fromHsv(hsv);
+    if (!colorModel.equal(newColor, cache.current)) {
+      cache.current = newColor;
+      onChangeRef(newColor);
```

Comment:
It seems to me that `onChangeRef` isn't the best name we could use here.
I mean, when most of the react-developers see a variable name that ends with `...Ref` they most probably expect that they have to add `.current`.

I would try something like `onChangeCallback(...)`. What do you think?

Reply:
Nice name for var, renamed it )